packages/transaction-controller/src/TransactionController.ts

```diff
@@ -1479,35 +1482,148 @@ export class TransactionController extends BaseController<
-      transactionMeta.custodyStatus = custodyStatus;
-    }
+    this.updateTransaction(
+      updatedTransactionMeta,
+      `TransactionController:updateCustodialTransaction - Custodial transaction updated`,
+    );
+  }
 
-    if (hash) {
-      transactionMeta.hash = hash;
+  /**
+   * Creates approvals for all unapproved transactions persisted.
+   */
+  initApprovals() {
+    const chainId = this.getChainId();
+    const unapprovedTxs = this.state.transactions.filter(
+      (transaction) =>
+        transaction.status === TransactionStatus.unapproved &&
+        transaction.chainId === chainId,
+    );
+
+    for (const txMeta of unapprovedTxs) {
+      this.processApproval(txMeta, {
+        shouldShowRequest: false,
+      }).catch((error) => {
+        if (error?.code === errorCodes.provider.userRejectedRequest) {
+          return;
+        }
+        console.error('Error during persisted transaction approval', error);
+      });
     }
+  }
 
-    this.updateTransaction(
-      transactionMeta,
-      `TransactionController:updateCustodialTransaction - Custodial transaction updated`,
+  /**
+   * Search transaction metadata for matching entries.
+   *
+   * @param opts - Options bag.
+   * @param opts.searchCriteria - An object containing values or functions for transaction properties to filter transactions with.
+   * @param opts.initialList - The transactions to search. Defaults to the current state.
+   * @param opts.filterToCurrentNetwork - Whether to filter the results to the current network. Defaults to true.
+   * @param opts.limit - The maximum number of transactions to return. No limit by default.
+   * @returns An array of transactions matching the provided options.
+   */
+  getTransactions({
+    searchCriteria = {},
+    initialList,
+    filterToCurrentNetwork = true,
+    limit,
+  }: {
+    searchCriteria?: any;
+    initialList?: TransactionMeta[];
+    filterToCurrentNetwork?: boolean;
+    limit?: number;
+  } = {}) {
+    const chainId = this.getChainId();
+    // searchCriteria is an object that might have values that aren't predicate
+    // methods. When providing any other value type (string, number, etc), we
+    // consider this shorthand for "check the value at key for strict equality
+    // with the provided value". To conform this object to be only methods, we
+    // mapValues (lodash) such that every value on the object is a method that
+    // returns a boolean.
+    const predicateMethods = mapValues(searchCriteria, (predicate) => {
+      return typeof predicate === 'function'
+        ? predicate
+        : (v: any) => v === predicate;
+    });
+
+    // If an initial list is provided we need to change it back into an object
+    // first, so that it matches the shape of our state. This is done by the
+    // lodash keyBy method. This is the edge case for this method, typically
+    // initialList will be undefined.
+    const transactionsToFilter = initialList
+      ? keyBy(initialList, 'id')
```

Comment:
Would it make more sense to leave initialList as an array and instead convert `this.state.transactions` to an array?

```typescript
const transactionsToFilter = initialList ?? Object.values(this.state.transactions);
```